### PR TITLE
Type descriptors for C#

### LINKED ISSUE #610: Compiled code initializes datatype fields with wrong default values
[cf #674]

In the following code, Dafny can prove that the variables `x`, `a`, and `b` are all greater than 0:

```
type pos = i:int | i > 0 witness 1
 
method m()
{
  var x:pos;
  assert x > 0;
  var p:(pos, pos);
  var (a, b) := p;
  assert a > 0;
  assert b > 0;
}
```

However, in the generated C# code, `a` and `b` are 0, which would violate the assertions in the Dafny code:

```
public static void m()
{
  BigInteger _22_x = pos.Witness;
  { }
  _System.Tuple2<BigInteger,BigInteger> _23_p = _System.Tuple2<BigInteger,BigInteger>.Default;
  _System.Tuple2<BigInteger,BigInteger> _let_tmp_rhs0 = _23_p;
  BigInteger _24_a = ((_System.Tuple2<BigInteger,BigInteger>)_let_tmp_rhs0)._0;
  BigInteger _25_b = ((_System.Tuple2<BigInteger,BigInteger>)_let_tmp_rhs0)._1;
  { }
  { }
}
```

### LINKED ISSUE #674: Incorrect default values for generic types
[cf #610]

In the code below, we should (provably) never reach the print statements, and yet the executable produced via C# does print them both (at least as of c580b1e8100fbb2a).  
[ Edit: As of 8/5, master (945ca6d989e0a5d9b6f67a86470ca4994fa0f9fb) exhibits correct behavior for js and go, incorrect behavior for java and C#)]

In the tuple portion of the example below, the emitted C# code is:
```
_System.Tuple2<BigInteger,BigInteger> _15_p = _System.Tuple2<BigInteger,BigInteger>.Default;
```
whereas we want the right-hand side at least to be `_System.Tuple2<pos, pos>.Default`.

The relevant line (for C#) appears to be this one:
https://github.com/dafny-lang/dafny/blob/ba5777b4faf18ec2dd0512490b55fa468e238f63/Source/Dafny/Compiler-Csharp.cs#L1056
where it’s calling `TypeNames` on the `TypeArgs`, which calls `TypeName`, which starts by calling `NormalizeExpand`, which converts `pos` into `int`. 

I'm afraid I'm not sure what the appropriate fix is, unless it's a version of `TypeName` (and hence `TypeNames` that takes an option to not expand the type it's handed.

```
type pos = i:int | i > 0 witness 1

datatype Generic<T> = Gen(p:T)

method Main() {
  var p:(pos, pos);
  var (a, b) := p;
  assert a > 0;
  assert b > 0;

  if a == 0 {
    assert false;
    print "We should not arrive here, and yet we do.\n";
  }

  var e:Generic<pos>;
  var Gen(n) := e;
  assert n > 0;
  if n == 0 {
    assert false;
    print "We should not arrive here, and yet we do.\n";
  }
}
```

### PR DESCRIPTION
**Pull request by RustanLeino**
Fixes #610
Fixes #674 

The compilers for Java, JavaScript, and Go use type descriptors for auto-init types. These are used to fabricate default values for type parameters. C# had previously tried to use reflection for this purpose, but that does not work for tuples and other datatypes that themselves take type parameters. This PR changes the compilation to C# to also use type descriptors, making the treatment of type descriptors uniform for these 4 compilers (compilation to C++ currently does not auto initialization via type parameters).

The type-descriptor compilation scheme is now documented in `docs/Compilation/AutoInitialization.md`.

Along the way to supporting type descriptors, many other things were fixed and made more uniform as well, for both the type-descriptor and default-value methods of each type. Among these changes was making datatype `Default` method take values, not type descriptors, as parameters for C#; a future improvement is to do the same change for the other compilers.

Other improvements and bug fixes:

* Don't complain about traits as type arguments for non-variant type parameters (C# and Java), see issues #406 and #599. Now, the only place traits are not supported as type parameters and for co- or contra-variant type parameters in C# and Java; these will be fixed once the C# and Java compilers add support for variant type parameters of datatypes.

* fix: Eta expansion of instance members

* fix: Name-capture problems (in lambdas, quantifiers, set/map comprehensions, and in tail-recursive functions/methods)

* fix: Don't use type parameters in `instanceof` for Java

* fix: Array coercions in Java

* fix: Latent `nil` issues in Go

* fix: Don't ignore functions in datatype cyclicity checks. Some datatypes that were previously (incorrectly) admitted will now give errors.

* fix: Print `null` as "null" (not as "") for possibly-null references in Java

Some of the internal changes:

* Easier customization of type parameters and type descriptors in compilers. This is now done in the compiler-specific `TypeArgDescriptorUse` method, see `docs/Compilation/AutoInitialization.md`.

* Handle both type parameters and type descriptors through the `TypeArgumentInstantiation`.

* refactor: Add `constructTypeParameterDefaultsFromTypeDescriptors` parameter to `DefaultValue`, `TypeInitialization`, and `TypeInitializationValue`. As a future improvement, the effect of the `inAutoInitContext` parameter to these methods is probably better refactored into separate methods.

* Resolve type parameters of tuple types and array types

* Refactor `CreateFunctionOrGetter`, to more easily switch to a function when a "getter" would need type-descriptor parameters.

* For Java, rename `dafny.Type` to `dafny.TypeDescriptor`, and `_type` to `_typeDescriptor`.

### CODE REVIEW

File: Source/Dafny/Compiler-java.cs [hunk 1]
```diff
@@ -766,47 +773,26 @@ protected class ClassWriter : IClassWriter {
...
+      Contract.Assume(variance.Count == typeArgs.Count);
       string s = IdProtect(fullCompileName);
       if (typeArgs.Count != 0) {
-        if (typeArgs.Exists(ComplicatedTypeParameterForCompilation)) {
-          Error(tok, "compilation does not support trait types as a type parameter; consider introducing a ghost", wr);
+        for (var i = 0; i < typeArgs.Count; i++) {
+          var v = variance[i];
+          var ta = typeArgs[i];
+          if (ComplicatedTypeParameterForCompilation(v, ta)) {
+            Error(tok, "compilation does not support trait types as a type parameter (got '{0}'{1}); consider introducing a ghost", wr,
```
Comment (davidcok) [inline Source/Dafny/Compiler-java.cs:787]: I thought traits as type parameters were allowed now.

File: Source/Dafny/Compiler-js.cs [hunk 2]
```diff
@@ -519,26 +519,31 @@ public JavaScriptCompiler(ErrorReporter reporter)
...
+        DeclareField("Witness", true, true, sst.Rhs, sst.tok, sw.ToString(), w);
+        d = TypeName_UDT(FullTypeName(udt), udt, wr, udt.tok) + ".Witness";
+      } else {
+        d = TypeInitializationValue(udt, wr, sst.tok, false, false);
       }
       using (var wDefault = w.NewBlock("static get Default()")) {
-        var udt = UserDefinedType.FromTopLevelDecl(sst.tok, sst);
-        var d = TypeInitializationValue(udt, wr, sst.tok, false);
-        wDefault.WriteLine("return {0};", d);
+        wDefault.WriteLine($"return {d};");
```
Comment (davidcok) [inline Source/Dafny/Compiler-js.cs:546]: Why bother with the using and wDefault temporary here. Just write 
w.NewBlock("static get Default()").WriteLine($"return {d};");

File: Source/Dafny/Compiler.cs [hunk 3]
```diff
@@ -211,28 +179,48 @@ protected interface IClassWriter {
...
       return tp.Characteristics.MustSupportZeroInitialization;
     }
 
-    protected virtual int EmitRuntimeTypeDescriptorsActuals(List<TypeArgumentInstantiation> typeArgs, Bpl.IToken tok, bool useAllTypeArgs, TargetWriter wr) {
+    protected abstract string TypeDescriptor(Type type, TextWriter wr, Bpl.IToken tok);
+
+    protected void EmitTypeDescriptorsActuals(List<TypeArgumentInstantiation> typeArgs, Bpl.IToken tok, TargetWriter wr, bool useAllTypeArgs = false) {
+      var prefix = "";
+      EmitTypeDescriptorsActuals(typeArgs, tok, wr, ref prefix, useAllTypeArgs);
+    }
```
Comment (davidcok) [inline Source/Dafny/Compiler.cs:208]: The method above could be used more often that it is. I recall seeing a number of instances of sep = ""; and then a call of EmitTypeDescriptorsActuals

File: Source/Dafny/Compiler.cs [hunk 4]
```diff
@@ -211,28 +179,48 @@ protected interface IClassWriter {
...
+
+    protected void EmitTypeDescriptorsActuals(List<TypeArgumentInstantiation> typeArgs, Bpl.IToken tok, TargetWriter wr, ref string prefix, bool useAllTypeArgs = false) {
```
Comment (davidcok) [inline Source/Dafny/Compiler.cs:210]: This method's task is mostly to emit the list, but has a side effect of setting prefix. This seems a real hack -- and is the ref parameter ever actually used after this call?

File: Source/DafnyRuntime/DafnyRuntime.cs [hunk 5]
```diff
@@ -1183,28 +1202,24 @@ public partial class Helpers {
...
-      System.Reflection.MethodInfo mInfo = ty.GetMethod("_DafnyDefaultValue");
-      if (mInfo != null) {
-        G g = (G)mInfo.Invoke(null, null);
-        return g;
-      } else {
-        return default(G);
-      }
+
+    public static readonly TypeDescriptor<bool> BOOL = new TypeDescriptor<bool>(false);
+    public static readonly TypeDescriptor<char> CHAR = new TypeDescriptor<char>('D');
```
Comment (davidcok) [inline Source/DafnyRuntime/DafnyRuntime.cs:1207]: This character 'D' as the default is used elsewhere in the code -- is it important that they all be the same character? in any case, should they all refer to one declaration?

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/TypeDescriptor.java [hunk 6]
```diff
@@ -118,28 +118,28 @@ public String toString() {
...
-    public static final Type<Integer> INT = new IntType(0);
-    public static final Type<Long> LONG = new LongType(0L);
-    public static final Type<Boolean> BOOLEAN = new BooleanType(Boolean.FALSE);
-    public static final Type<Character> CHAR = new CharType('D');
+    public static final TypeDescriptor<Byte> BYTE = new ByteType((byte)0);
+    public static final TypeDescriptor<Short> SHORT = new ShortType((short)0);
+    public static final TypeDescriptor<Integer> INT = new IntType(0);
+    public static final TypeDescriptor<Long> LONG = new LongType(0L);
+    public static final TypeDescriptor<Boolean> BOOLEAN = new BooleanType(Boolean.FALSE);
+    public static final TypeDescriptor<Character> CHAR = new CharType('D');
```
Comment (davidcok) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/TypeDescriptor.java:126]: Another instance of 'D'

File: Test/dafny0/SubsetTypes.dfy.expect [hunk 7]
```diff
@@ -391,5 +391,13 @@ Execution trace:
 SubsetTypes.dfy(455,9): Error: value does not satisfy the subset constraints of 'nat'
 Execution trace:
     (0,0): anon0
+SubsetTypes.dfy(465,7): Error: cannot find witness that shows type is inhabited (only tried 'D'); try giving a hint through a 'witness' or 'ghost witness' clause
```
Comment (davidcok) [inline Test/dafny0/SubsetTypes.dfy.expect:394]: SubsetTypes.dfy failed CI -- is it still a problem or is the expect file obsolete?

File: docs/Compilation/AutoInitialization.md [hunk 8]
```diff
@@ -0,0 +1,534 @@
+<p></p> <!-- avoids duplicate title -->
+
+Automatic Initialization of Variables
```
Comment (davidcok) [inline docs/Compilation/AutoInitialization.md:3]: Should I adapt this document for the RM as well? Or just refer to it? Or have the RM replace this document?

File: docs/Compilation/AutoInitialization.md [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+```
+public static TypeDescriptor<T> NULL<T>() where T : class {
+  return new TypeDescriptor<T>(null);
+}
+```
+
+## Collection types
+
+Type descriptors for collection types are provided by the `_TypeDescriptor()` method in
+the type's companion class. In each case, the return returned is a value computed once
```
Comment (davidcok) [inline docs/Compilation/AutoInitialization.md:354]: perhaps 'the quantity returned'

File: docs/Compilation/AutoInitialization.md [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
+The companion class of a subset type `S<TT>` contains a method `_TypeDescriptor` that returns
+a type descriptor for `S<TT>`.
+
+```
+public static Dafny.TypeDescriptor<B> _TypeDescriptor(TypeDescriptor<T> td_T, ...) {
+  return new Dafny.TypeDescriptor<B>(Dve);
+}
+```
+
+where the list of type parameters denoted by `T, ...` are the auto-init type parameters from `TT`,
```
Comment (davidcok) [inline docs/Compilation/AutoInitialization.md:393]: are -> is

File: Source/Dafny/Compiler-Csharp.cs [hunk 11]
```diff
@@ -760,26 +842,23 @@ protected class ClassWriter : IClassWriter
...
-        IdName(m));
-      if (m.TypeArgs.Count != 0) {
-        wr.Write("<{0}>", TypeParameters(m.TypeArgs));
-      }
-      wr.Write("(");
-      int nIns;
+        IdName(m),
+        TypeParameters(TypeArgumentInstantiation.ToFormals(ForTypeParameters(typeArgs, m, lookasideBody))));
+      var sep = "";
+      var nIns = WriteRuntimeTypeDescriptorsFormals(m, ForTypeDescriptors(typeArgs, m, lookasideBody), wr, ref sep, tp => $"{DafnyTypeDescriptor}<{tp.CompileName}> {FormatTypeDescriptorVariable(tp)}");
```
Comment (davidcok) [inline Source/Dafny/Compiler-Csharp.cs:853]: Is there a reason we Write...Formals bout Emit...Actuals (and Emit most other things)?

File: Source/Dafny/Compiler-java.cs (hunk 1 above)
Comment (RustanLeino) [reply Source/Dafny/Compiler-java.cs:787]: In the language, yes. But they are not fully supported in the C# and Java compilers, if the type parameter is co- or contra-variant. This requires support of co-variant datatypes in C# and Java, which is still on the to-do list.

**Review by davidcok — CHANGES_REQUESTED** [review]
Minor questions and comments for response

File: Source/Dafny/Compiler-js.cs (hunk 2 above)
Comment (RustanLeino) [reply Source/Dafny/Compiler-js.cs:546]: Sure

File: Source/Dafny/Compiler.cs (hunk 3 above)
Comment (RustanLeino) [reply Source/Dafny/Compiler.cs:208]: Apparently, all the other 16 uses make use of the `prefix` `ref`-parameter after the call.

File: Source/Dafny/Compiler.cs (hunk 4 above)
Comment (RustanLeino) [reply Source/Dafny/Compiler.cs:210]: Yes, it's it returned as given, if nothing was emitted, and it is returned as `", "` if something was emitted. This helps the caller with separating the type-descriptor actuals from further actual parameters.

File: Source/DafnyRuntime/DafnyRuntime.cs (hunk 5 above)
Comment (RustanLeino) [reply Source/DafnyRuntime/DafnyRuntime.cs:1207]: Good point. It's used in the various compilers and in the verifier, and they must all agree. I added a named constant to `CharType`, and added comments to `DafnyRuntime.{cs,js,go}` and `TypeDescriptor.java' (the C++ compiler does not yet support type descriptors).

Btw, the reason it's not `'\0'` is that `'\0'` is a nuisance to deal with in printed output (like the test suite).

File: docs/Compilation/AutoInitialization.md (hunk 8 above)
Comment (RustanLeino) [reply docs/Compilation/AutoInitialization.md:3]: Excellent.

File: Test/dafny0/SubsetTypes.dfy.expect (hunk 7 above)
Comment (RustanLeino) [reply Test/dafny0/SubsetTypes.dfy.expect:394]: I had forgotten to update the `.expect` file in the `allocated1/dafny0` folder. Thanks.

File: docs/Compilation/AutoInitialization.md (hunk 9 above)
Comment (RustanLeino) [reply docs/Compilation/AutoInitialization.md:354]: done

File: docs/Compilation/AutoInitialization.md (hunk 10 above)
Comment (RustanLeino) [reply docs/Compilation/AutoInitialization.md:393]: done

File: Source/Dafny/Compiler-Csharp.cs (hunk 11 above)
Comment (RustanLeino) [reply Source/Dafny/Compiler-Csharp.cs:853]: I think the `Emit...` methods started off as ones that were overridden by the individual compilers, whereas `Write...` methods were (non-overridable) worker routines. I'd be surprised if this "rule" is applied consistently. I will leave this unchanged for now, hoping for a future refactoring pass to clean up and apply consistency.

**Review by davidcok — APPROVED** [review]

**Pull request merged**